Firestore/core/src/firebase/firestore/core/query.cc

```diff
@@ -247,6 +250,68 @@ bool Query::MatchesBounds(const Document& doc) const {
   return true;
 }
 
+model::DocumentComparator Query::Comparator() const {
+  OrderByList ordering = order_bys();
+
+  bool has_key_ordering = false;
+  for (const OrderBy& order_by : ordering) {
+    if (order_by.field() == FieldPath::KeyFieldPath()) {
+      has_key_ordering = true;
+      break;
+    }
+  }
+  HARD_ASSERT(has_key_ordering,
+              "QueryComparator needs to have a key ordering.");
+
+  return DocumentComparator(
+      [ordering](const Document& doc1, const Document& doc2) {
+        for (const OrderBy& order_by : ordering) {
+          ComparisonResult comp = order_by.Compare(doc1, doc2);
+          if (!util::Same(comp)) return comp;
+        }
+        return ComparisonResult::Same;
+      });
+}
+
+const std::string& Query::CanonicalId() const {
+  if (!canonical_id_.empty()) return canonical_id_;
+
+  std::string result;
+  absl::StrAppend(&result, path_.CanonicalString());
+
+  if (collection_group_) {
+    absl::StrAppend(&result, "|cg:", *collection_group_);
+  }
+
+  // Add filters.
+  absl::StrAppend(&result, "|f:");
```

Comment:
Question: in the canonical representation, filters and order by's get a tag even if the corresponding list is empty, whereas all other elements are not added at all if they are not present in the query. Is there any particular reason for this format?

Reply:
I have no idea, unfortunately. While porting this I found it to be idiosyncratic myself.

Changing it will either requires us to write a migration or causes nearly all queries to become lost, neither of which is appealing given that this is just a minor annoyance.